File: src/main/java/io/gitlab/intended/storagenetworks/block/entity/MasterBlockEntity.java
package io.gitlab.intended.storagenetworks.block.entity;

import io.gitlab.intended.storagenetworks.registry.BlockEntityTypeRegistry;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.Tag;

import java.util.ArrayList;
import java.util.List;

public class MasterBlockEntity extends BlockEntity {

    private ListTag storagePos = new ListTag();
    private List<int[]> storagePosArray = new ArrayList<>();

    public MasterBlockEntity() {
        super(BlockEntityTypeRegistry.MASTER);
    }

    @Override
    public CompoundTag toTag(CompoundTag tag) {
        super.toTag(tag);

        storagePosArray.forEach(storage -> {
            CompoundTag compoundTag = new CompoundTag();
            compoundTag.putInt("x", storage[0]);
            compoundTag.putInt("y", storage[1]);
            compoundTag.putInt("z", storage[2]);
            storagePos.add(compoundTag);
        });

        tag.put("storagePos", storagePos);

        return tag;
    }

    @Override
    public void fromTag(CompoundTag tag) {
        super.fromTag(tag);

        storagePos = (ListTag) tag.get("storagePos");
        storagePosArray.clear();
        if (storagePos != null) {
            for (Tag storage : storagePos) {
                CompoundTag compoundTag = (CompoundTag) storage;
                int[] ints = {
                        compoundTag.getInt("x"),
                        compoundTag.getInt("y"),
                        compoundTag.getInt("z")
                };
                storagePosArray.add(ints);
            }
        }
    }

}

File: src/main/kotlin/io/gitlab/intended/storagenetworks/block/entity/MasterBlockEntity.kt
package io.gitlab.intended.storagenetworks.block.entity

import io.gitlab.intended.storagenetworks.block.LinkCableBlock
import io.gitlab.intended.storagenetworks.block.entity.type.BlockEntityTypeRegistry
import io.gitlab.intended.storagenetworks.tag2Pos
import net.fabricmc.fabric.api.util.NbtType
import net.minecraft.block.entity.BlockEntity
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.world.World

class MasterBlockEntity : BlockEntity(BlockEntityTypeRegistry.MASTER) {

    private var linkCables = ListTag()

    private fun validateConnectors(world: World) {
        val linkCableSet = HashSet<CompoundTag>()
        linkCables.forEach { linkCableTag ->
            linkCableTag as CompoundTag
            val linkCablePos = tag2Pos(linkCableTag)
            val linkCableBlock = world.getBlockState(linkCablePos).block
            if (linkCableBlock is LinkCableBlock) {
                val linkCableNbt = world.getBlockEntity(linkCablePos)!!.toTag(CompoundTag())

                val linkCableHasMaster = linkCableNbt.getBoolean("hasMaster")
                val linkCableMasterPos = tag2Pos(linkCableNbt.getCompound("masterPos"))

                if (linkCableHasMaster and (linkCableMasterPos == pos)) {
                    linkCableSet.add(linkCableTag)
                }
            }
        }
        linkCables.clear()
        linkCables.addAll(linkCableSet)
    }

    override fun toTag(tag: CompoundTag): CompoundTag {
        super.toTag(tag)

        tag.put("linkCables", linkCables)

        return tag
    }

    override fun fromTag(tag: CompoundTag) {
        super.fromTag(tag)

        linkCables = tag.getList("linkCables", NbtType.COMPOUND)
    }

    override fun markDirty() {
        if (world != null) validateConnectors(world!!)
        super.markDirty()
    }

}
